Expensify/App
# Local notifications on iOS and Android

**Note:** As of now this has been tested and is working well on iOS. Still need to test on Android.

### Details
This PR introduces a local notification library for iOS and Android. It will display a locally-generated push notification whenever a local notification is generated on web/desktop, so:

0. To run locally, I had to:

```bash
# This first command will reset the cache then not do anything
react-native start --reset-cache 
npm run ios
```

1. When the app is in the foreground and you receive a chat notification from a chat other than the main chat.
2. When the app is in the background but is still connected via Pusher.
    - Caveat: I'm not sure the rules for how long the app will stay connected to Pusher when it is backgrounded. During my testing, I've found some puzzling behavior...
        1. Immediately after backgrounding the app, send a message.
        1. The message is received by Pusher and a local notification displays, adds the message to Onyx, can be tapped to view the report, etc.. (new)
        1. Whenever Pusher disconnects (I still haven't figured out how to consistently know if this has happened, at least not on dev), send another message.
        1. After pusher is disconnected (and the server knows), send another message. We'll send a UA push notification and the push notification is received by the client and displayed, but the callback does not seem to execute. Not sure if this is a regression from this branch, but it seems like it hasn't worked at any point in the past. I searched LogSearch and my local logs for [this line](https://github.com/Expensify/Expensify.cash/blob/a40503a38a18b796adff02fb90d2b3684ce580b8/src/libs/actions/Report.js#L719) and it seems it's never logged. So it doesn't add the message to Onyx ❌ . However, the notification can be tapped, and it _does_ seem to open the correct report (on iOS at least). When that report is opened, it reconnects to pusher and reloads messages. This is bad for offline-first.
        1. There might be some time between when pusher is connected and when it's not, during which push notifications aren't received? But they're being sent, so maybe that's just flakiness/delays with UA?
            - Sometimes, when a message like this happens and no notification displays, then you later receive + tap on a UA notification, the local notification will display afterwords?

    - TODO: discovered a bug where, if you receive push UA push notifications and they're displayed, then you tap to open that chat, then background the app, those message still appear as unread per the unread counter
    
**Update:** I think I understand the scenario in which notifications do not display in the background. This will happen when the client disconnects from the Pusher channel (happens after some undetermined amount of time with the app backgrounded), but the server still thinks they're connected. Need to dig into how this can happen and how to fix it.

### Fixed Issues
Fixes (partial) https://github.com/Expensify/Expensify/issues/158830

Note that this doesn't improve _push notifications_ in any other states (and in fact, still needs to be tested to make sure it doesn't break UA push notifications, which it may well do...) But assuming that it doesn't break UA push notifications, it will be a substantial improvement to the notification experience on iOS/Android.

### Tests / QA Steps
1. Sign into E.cash on iOS/Android in one account (A), and another account (B) on another platform.
1. Open a chat with A and B.
1. Send a chat from B to A. The new message should appear and there should be no notification.
1. Open a different chat in account A.
1. Send a message from B to A. A notification should appear for A Do not tap the notification. It should disappear after a few seconds.
1. (iOS) Send a message from B to A. A notification should appear for A. Swipe the notification up to dismiss it. It should disappear and nothing else should happen.
1. Send a message from B to A. A notification should appear for A. Tap the notification. The chat between A and B should open, and all three messages should immediately be present.
1. As A, background the app.
1. Send a message from B to A. A notification should appear for A. The app's unread counter badge should increment. Do not tap the notification. It should disappear after a few seconds.
1. (iOS) Send a message from B to A. A notification should appear for A. The app's unread counter badge should increment Swipe the notification up to dismiss it. It should disappear and nothing else should happen.
1. Send a message from B to A. A notification should appear for A. Tap the notification. The chat between A and B should open, and all three messages should immediately be present.

### Tested On


### Screenshots
#### Web


#### Mobile Web


#### Desktop


#### iOS


#### Android
